### ISSUE TITLE
pojo mapper ignores setter for different type

### ISSUE THREAD

**Issue by rainoko** [jOOQ/jOOQ#5796]
In versions 3.7 and 3.9. 
If pojo has field 

```java
Set<Long> f
```

And setter 

```
public setF(Long[])
```

Then when db has field as array jooq ignores suitable setter and failes to map array to set field.
We do not use jpa annotations.

This worked in 3.5

**Comment by rainoko** [comment]
Db postgresql 9.3

**Comment by lukaseder** [comment]
Thank you very much for reporting. I'm sorry I'm not quite sure what you mean. Could you please provide an example to help reproduce this issue? (including the relevant DDL and your code generator configuration)

**Comment by rainoko** [comment]
made some example app with spring boot and gradle
https://github.com/rainoko/JooqArray

I used h2 for example. The example works with 3.5.3 but not 3.9.0

```
CREATE TABLE IF NOT EXISTS TESTTAB (
  id                      INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
  dataArr                 ARRAY
);

```
Code
```
create.insertInto(Testtab.TESTTAB).set(Testtab.TESTTAB.DATAARR, new Integer[]{1, 2}).execute();
    Domain d = create.select().from(Testtab.TESTTAB).fetchOne().into(Domain.class);
    System.out.println(d.getDataarr());

...........
public class Domain {

  private Long id;
  private List<Integer> dataarr = new ArrayList<>();

  public Long getId() {
    return id;
  }

  public void setId(Long id) {
    this.id = id;
  }

  public List<Integer> getDataarr() {
    return dataarr;
  }

  public void setDataarr(Integer[] in) {
//    dataArr = in;
    Arrays.stream(in).forEach(i -> dataarr.add((Integer)i));
  }
}
```

So jooq cant map field in 3.5.3 jooq identifies setter for dataarr. But in 3.9.0 jooq wont choose setter but tries to map with field with error
`org.jooq.exception.DataTypeException: Cannot convert from 1 (class java.lang.Integer) to interface java.util.List`

**Comment by rainoko** [comment]
3.5.3 identifies mapping POJO fields by type.getFields which ignores non public fields. So jooq falls back to setters.
3.9.0 (this is actually earlier modification) Utils.getInstanceMembers also uses getDeclaredFields and then it finds private field which is suitable by name but in my case not suitable by type - jooq never did eny type check for fields.

I dont know the historical reason but why not first use method and after that field - okey reason might be how to find suitable setter if there are more than one.

Anyway best way should be the type check of field also. But this is not easy refactoring. Right now I have to rewrite my class to have array fields and getters to return Collection if needed.

**Comment by lukaseder** [comment]
Thanks for the explanation. So, the problem stems from the fact that you don't strictly follow the JavaBeans convention.

I'm not sure what change triggered the regression here, but the `DefaultRecordMapper` will always use setters by name, not by type. If you overload setters (or use a setter with a "wrong" type), then `DefaultRecordMapper` will call that setter nonetheless.

In any case, this will need further investigation. Thanks again for reporting. I'll comment on this again, soon.

**Comment by rainoko** [comment]
setter is correct for jooq. problem is that field itself is not compatible.
postgre array cant be mapped to List/or Set.
maybe jooq should map 
1. try constructor
2. try setters
3. try fields
4. try declared fields

right now i resolve problem with writing acceptable constructor. so it doesnt seems to be any important bug. Maybe its not bug but feature

**Comment by lukaseder** [comment]
Regarding the algorithm, it is clearly specified in [`DefaultRecordMapper`](https://www.jooq.org/javadoc/latest/org/jooq/impl/DefaultRecordMapper.html). Sure, we can debate the algorithm, but no breaking change will be made in a minor release as many jOOQ users rely on the current behaviour (which includes setting both  the matching setter(s) *and* the fields).

Within the limits of the current specification, you have the following possible workarounds, however:

- Rename your `dataarr` field to something else. Then, `DefaultRecordMapper` will ignore it
- Make your `dataarr` field `final`

In any case, I do agree that jOOQ should be able to convert arrays to lists. There's a pending feature request for this: https://github.com/jOOQ/jOOQ/issues/3062

I will close this issue as a duplicate for #3062. If you're interested in discussing the mapping algorithm as a whole, I would like to invite you to create a new issue dedicated for that (or send a mail to the jOOQ user group).

**Comment by lukaseder** [comment]
This issue is effectively fixed now with #3062 in jOOQ 3.10.
